### PR TITLE
added exception groups

### PR DESCRIPTION
**Pull request by iritkatriel**
First step of PEP-654: added Exception Groups (without updating traceback formatting).

### CODE REVIEW

File: Include/pyerrors.h [hunk 1]
```diff
@@ -65,6 +65,8 @@ PyAPI_FUNC(const char *) PyExceptionClass_Name(PyObject *);
 
 PyAPI_DATA(PyObject *) PyExc_BaseException;
 PyAPI_DATA(PyObject *) PyExc_Exception;
+PyAPI_DATA(PyObject *) PyExc_BaseExceptionGroup;
+PyAPI_DATA(PyObject *) PyExc_ExceptionGroup;
```
Comment (iritkatriel) [inline Include/pyerrors.h:69]: I don't know what this is for, assuming it's needed.

File: Lib/test/exception_hierarchy.txt [hunk 2]
```diff
@@ -2,7 +2,9 @@ BaseException
  +-- SystemExit
  +-- KeyboardInterrupt
  +-- GeneratorExit
+ +-- BaseExceptionGroup
  +-- Exception
+      +-- ExceptionGroup
```
Comment (iritkatriel) [inline Lib/test/exception_hierarchy.txt]: Unfortunately we can't express here that ExceptionGroup subclasses BaseExceptionGroup as well.

File: PC/python3dll.c [hunk 3]
```diff
@@ -722,6 +722,7 @@ EXPORT_DATA(PyExc_ArithmeticError)
 EXPORT_DATA(PyExc_AssertionError)
 EXPORT_DATA(PyExc_AttributeError)
 EXPORT_DATA(PyExc_BaseException)
+EXPORT_DATA(PyExc_BaseExceptionGroup)
```
Comment (iritkatriel) [inline PC/python3dll.c:756]: I don't know what this is for, assuming it's needed.

File: Include/pyerrors.h (hunk 1 above)
Comment (gvanrossum) [reply Include/pyerrors.h:69]: You mean the funky macro around the return type? It’s to make dynamic loaders publish the symbol on all platforms (but mostly Windows IIRC).

File: Lib/test/exception_hierarchy.txt (hunk 2 above)
Comment (gvanrossum) [reply Lib/test/exception_hierarchy.txt]: Add a note in parentheses?

**Review by gvanrossum — COMMENTED** [review]
Is this what we will merge once the PEP is accepted?

(Note that the master -> main rename will happen at that time too. But supposedly PSs will be automatically adjusted. We’ll see.)

**Comment by gvanrossum** [comment]
Probably mark as Draft?

File: Lib/test/exception_hierarchy.txt (hunk 2 above)
Comment (iritkatriel) [reply Lib/test/exception_hierarchy.txt]: There seems to be an unused feature that interprets whatever is in "(..)" as a platform name and then ignores any exception where the platform is not the current platform (so it won't break the test but will weaken it).

But "[...]" seems ok - anything in them appears to be ignored, and when I changed ExceptionGroup to something invalid the test failed.


https://github.com/python/cpython/blob/8bbfeb3330c10d52274bb85fce59ae614f0500bf/Lib/test/test_baseexception.py#L49

File: Include/pyerrors.h (hunk 1 above)
Comment (iritkatriel) [reply Include/pyerrors.h:69]: Yes, I'm just pointing out that I copy-pasted it like a donkey.

**Review by iritkatriel — COMMENTED** [review]
> Is this what we will merge once the PEP is accepted?

That's the idea. The branch of the other PR is very old and compile.c/eval.c changed beyond what can be fixed with a rebase. So I started from scratch on a recent cut of master.  Marked as draft for now.

**Comment by gvanrossum** [comment]
You should probably merge this on main.

**Comment by iritkatriel** [comment]
I rebased and regenerated Python/importlib_external.h

**Comment by iritkatriel** [comment]
Superseded by #28569.

**Pull request closed without merging**